README.md
# blurhash-to-css

> Converts a [BlurHash](https://blurha.sh) to a [CSS Object](https://developer.mozilla.org/en-US/docs/Web/API/ElementCSSInlineStyle/style) using [TypeScript](https://www.typescriptlang.org/), [Rust](https://www.rust-lang.org/), and [WebAssembly](https://github.com/rustwasm/wasm-pack).

[![NPM version](http://img.shields.io/npm/v/blurhash-to-css.svg?style=flat-square)](https://www.npmjs.com/package/blurhash-to-css) [![NPM downloads](http://img.shields.io/npm/dm/blurhash-to-css.svg?style=flat-square)](https://www.npmjs.com/package/blurhash-to-css) [![Follow JamieMason on GitHub](https://img.shields.io/github/followers/JamieMason.svg?style=social&label=Follow)](https://github.com/JamieMason) [![Follow fold_left on Twitter](https://img.shields.io/twitter/follow/fold_left.svg?style=social&label=Follow)](https://twitter.com/fold_left)

## 🌩 Installation

```
npm install blurhash-to-css
```

## 📣 Summary

[BlurHash](https://blurha.sh) is a compact representation of a placeholder for an image, which makes it a great option to [improve perceived load times](https://blog.imgix.com/2021/01/26/blurhash). But, a BlurHash is rendered using a `<canvas>` element, which means client-side JavaScript is needed on the critical path when loading our application.

This tool creates the same visual effect using CSS, which can be rendered on the Server without any dependency on client-side JavaScript.

An example can be seen at https://blurhash-to-css.vercel.app and the source is in this repo at [`/example`](https://github.com/JamieMason/blurhash-to-css/tree/main/example).

## 🤝 Credit

- The image buffer to CSS conversion is a port to Rust of [this Script](https://github.com/joe-bell/plaiceholder/blob/d09df807df5d11c5d91a7971c2f90e2faa845843/packages/css/src/index.ts#L18-L51) originally by [**@joe-bell**](https://github.com/joe-bell) for [plaiceholder: Beautiful image placeholders, without the hassle](https://plaiceholder.co/).
- A lot of help came from the source code in [wasm-image](https://github.com/peerigon/wasm-image) by [**@acidicX**](https://github.com/acidicX).

## 🕹 Usage

```ts
import { blurhashToCss } from 'blurhash-to-css';

const css = blurhashToCss('eCF6B#-:0JInxr?@s;nmIoWUIko1%NocRk.8xbIUaxR*^+s;RiWAWU');
```

↓↓↓↓↓

```json
{
  "backgroundImage": "linear-gradient(90deg, rgb(139,153,96) 10%,rgb(153,160,118) 10% 20%,rgb(170,172,142) 20% 30%,rgb(177,180,151) 30% 40%,rgb(174,180,146) 40% 50%,rgb(162,172,130) 50% 60%,rgb(148,162,114) 60% 70%,rgb(136,150,104) 70% 80%,rgb(131,145,95) 80% 90%,rgb(130,144,91) 90% 100%),linear-gradient(90deg, rgb(109,129,65) 10%,rgb(117,131,80) 10% 20%,rgb(130,138,101) 20% 30%,rgb(142,148,113) 30% 40%,rgb(146,150,114) 40% 50%,rgb(136,144,103) 50% 60%,rgb(116,130,88) 60% 70%,rgb(102,119,75) 70% 80%,rgb(104,118,71) 80% 90%,rgb(112,123,72) 90% 100%),linear-gradient(90deg, rgb(110,122,70) 10%,rgb(114,122,85) 10% 20%,rgb(126,129,106) 20% 30%,rgb(143,140,122) 30% 40%,rgb(151,147,128) 40% 50%,rgb(144,142,122) 50% 60%,rgb(123,126,107) 60% 70%,rgb(106,111,90) 70% 80%,rgb(108,109,83) 80% 90%,rgb(118,117,83) 90% 100%),linear-gradient(90deg, rgb(134,133,102) 10%,rgb(141,138,120) 10% 20%,rgb(155,150,144) 20% 30%,rgb(172,164,161) 30% 40%,rgb(179,173,168) 40% 50%,rgb(173,168,162) 50% 60%,rgb(156,151,146) 60% 70%,rgb(137,132,127) 70% 80%,rgb(127,124,110) 80% 90%,rgb(128,124,101) 90% 100%),linear-gradient(90deg, rgb(116,124,95) 10%,rgb(125,129,112) 10% 20%,rgb(143,141,135) 20% 30%,rgb(160,156,153) 30% 40%,rgb(169,164,160) 40% 50%,rgb(165,161,155) 50% 60%,rgb(148,145,141) 60% 70%,rgb(124,125,118) 70% 80%,rgb(107,110,97) 80% 90%,rgb(99,106,82) 90% 100%),linear-gradient(90deg, rgb(95,105,51) 10%,rgb(99,105,66) 10% 20%,rgb(110,111,88) 20% 30%,rgb(125,122,107) 30% 40%,rgb(136,131,115) 40% 50%,rgb(134,129,111) 50% 60%,rgb(120,116,98) 60% 70%,rgb(97,97,77) 70% 80%,rgb(78,84,51) 80% 90%,rgb(70,78,30) 90% 100%),linear-gradient(90deg, rgb(109,106,21) 10%,rgb(109,104,44) 10% 20%,rgb(112,104,67) 20% 30%,rgb(122,109,83) 30% 40%,rgb(131,114,88) 40% 50%,rgb(130,112,83) 50% 60%,rgb(119,103,70) 60% 70%,rgb(104,91,51) 70% 80%,rgb(93,84,34) 80% 90%,rgb(92,84,24) 90% 100%)",
  "backgroundPosition": "0 0 ,0 16.666666666666664%,0 33.33333333333333%,0 50%,0 66.66666666666666%,0 83.33333333333334%,0 100%",
  "backgroundSize": "100% 14.285714285714286%",
  "backgroundRepeat": "no-repeat",
  "filter": "blur(24px)",
  "transform": "scale(1.2)"
}
```

To use this a CommonJS project, you should destructure it like so:

```ts
const { blurhashToCss } = require('blurhash-to-css');
```

## 🛠 Options

### `blurhash` (required)

A small string (or array of strings) generated to represent a blurry version of an image which will download sometime soon. Have a play around on [BlurHash](https://blurha.sh), [plaiceholder.co](https://plaiceholder.co/), and have a read of [BlurHash for imgix: An Alternative to Generic Image Placeholders](https://blog.imgix.com/2021/01/26/blurhash) to try them out and find out more.

### `width` and `height` (optional)

The BlurHash is decoded into a 10 x 10 image by default, you can make this larger and it will look more like the final image, but it will also generate a lot more CSS.

```ts
const css = blurhashToCss('LEHLh[WB2yk8pyoJadR*.7kCMdnj', {
  height: 30,
  width: 30,
});
```


CHANGELOG.md
## [0.5.5](https://github.com/JamieMason/blurhash-to-css/compare/0.4.3...0.5.5) (2022-11-12)

### Bug Fixes

- **cargo:** update dependencies
  ([3c6e1d7](https://github.com/JamieMason/blurhash-to-css/commit/3c6e1d7ee1b293da4659c97f5a69730e0ba931be))
- **npm:** update dependencies
  ([e43dec9](https://github.com/JamieMason/blurhash-to-css/commit/e43dec97bd4b5756a9cd8919c4f6ebcad9182e48))

### Features

- **blurhash:** accept arrays of blurhashes
  ([3afacfc](https://github.com/JamieMason/blurhash-to-css/commit/3afacfcc75e24461f9ba9ccbfeb61287eb550662))

## [0.4.3](https://github.com/JamieMason/blurhash-to-css/compare/0.3.3...0.4.3) (2021-02-08)

### Features

- **perf:** remove unneeded image buffer
  ([c3aadc9](https://github.com/JamieMason/blurhash-to-css/commit/c3aadc909ce15ff1260252c17c9a4e598598783a))

## [0.3.3](https://github.com/JamieMason/blurhash-to-css/compare/0.3.2...0.3.3) (2021-02-07)

### Bug Fixes

- **options:** make width and height optional
  ([0dbe68d](https://github.com/JamieMason/blurhash-to-css/commit/0dbe68d7363fdb84b6f60a3e02e562a782b6b65a))

## [0.3.2](https://github.com/JamieMason/blurhash-to-css/compare/0.3.1...0.3.2) (2021-02-07)

### Bug Fixes

- **node:** revert bundling and inlining source
  ([5f182ea](https://github.com/JamieMason/blurhash-to-css/commit/5f182ead880ddd57bed3416abaaca7a82726bf9e))

## [0.3.1](https://github.com/JamieMason/blurhash-to-css/compare/0.2.1...0.3.1) (2021-02-07)

### Features

- **node:** bundle source and inline wasm
  ([a3d9cfc](https://github.com/JamieMason/blurhash-to-css/commit/a3d9cfca12dd2d59ffc5b34ffdd54cb719a6c7f0))

## [0.2.1](https://github.com/JamieMason/blurhash-to-css/compare/0.1.0...0.2.1) (2021-02-06)

### Bug Fixes

- **types:** add filter and transform properties
  ([a475c8f](https://github.com/JamieMason/blurhash-to-css/commit/a475c8f9d1d4eb50e951321bb576295e72cde05e))

### Features

- **blurhash:** port to rust and webassembly
  ([d469b24](https://github.com/JamieMason/blurhash-to-css/commit/d469b24c10ed6d2ba20b06c9fb23c6e8acf2e2a3))

# [0.1.0](https://github.com/JamieMason/blurhash-to-css/compare/7b345388a47f0b3cddd3c31e8ff9ae21f74cd16e...0.1.0) (2021-02-04)

### Features

- **blurhash:** convert blurhash to css
  ([7b34538](https://github.com/JamieMason/blurhash-to-css/commit/7b345388a47f0b3cddd3c31e8ff9ae21f74cd16e))


src/index.spec.ts
/** More tests can be found in /native/src/lib.rs */
import { blurhashToCss } from '.';

it('converts a blurhash into a CSS object', () => {
  expect(
    blurhashToCss({
      blurhash: 'eCF6B#-:0JInxr?@s;nmIoWUIko1%NocRk.8xbIUaxR*^+s;RiWAWU',
      width: 400,
      height: 600,
    })
  ).toEqual({
    backgroundImage: [
      'linear-gradient(90deg,rgb(143,156,100) 10%,rgb(158,164,124) 10% 20%,rgb(174,176,146) 20% 30%,rgb(181,183,155) 30% 40%,rgb(177,182,148) 40% 50%,rgb(164,174,132) 50% 60%,rgb(150,163,116) 60% 70%,rgb(139,153,105) 70% 80%,rgb(133,148,98) 80% 90%,rgb(131,146,94) 90% 100%)',
      'linear-gradient(90deg,rgb(122,139,78) 10%,rgb(134,145,99) 10% 20%,rgb(148,154,119) 20% 30%,rgb(158,161,128) 30% 40%,rgb(156,162,124) 40% 50%,rgb(145,155,111) 50% 60%,rgb(128,142,95) 60% 70%,rgb(116,132,84) 70% 80%,rgb(115,130,80) 80% 90%,rgb(119,132,80) 90% 100%)',
      'linear-gradient(90deg,rgb(103,121,57) 10%,rgb(109,123,74) 10% 20%,rgb(122,130,93) 20% 30%,rgb(135,139,105) 30% 40%,rgb(140,143,107) 40% 50%,rgb(130,136,98) 50% 60%,rgb(109,121,82) 60% 70%,rgb(94,109,69) 70% 80%,rgb(99,110,67) 80% 90%,rgb(111,117,71) 90% 100%)',
      'linear-gradient(90deg,rgb(111,121,72) 10%,rgb(116,123,89) 10% 20%,rgb(129,131,109) 20% 30%,rgb(145,143,124) 30% 40%,rgb(152,149,129) 40% 50%,rgb(143,142,122) 50% 60%,rgb(123,125,107) 60% 70%,rgb(106,110,91) 70% 80%,rgb(109,111,84) 80% 90%,rgb(119,118,84) 90% 100%)',
      'linear-gradient(90deg,rgb(130,132,97) 10%,rgb(138,136,117) 10% 20%,rgb(153,147,141) 20% 30%,rgb(168,161,157) 30% 40%,rgb(175,168,161) 40% 50%,rgb(167,161,153) 50% 60%,rgb(148,144,137) 60% 70%,rgb(130,126,119) 70% 80%,rgb(125,122,105) 80% 90%,rgb(129,125,99) 90% 100%)',
      'linear-gradient(90deg,rgb(132,134,105) 10%,rgb(142,140,126) 10% 20%,rgb(158,153,150) 20% 30%,rgb(174,167,167) 30% 40%,rgb(180,174,171) 40% 50%,rgb(173,169,164) 50% 60%,rgb(155,151,147) 60% 70%,rgb(135,132,127) 70% 80%,rgb(123,122,108) 80% 90%,rgb(121,121,97) 90% 100%)',
      'linear-gradient(90deg,rgb(113,121,90) 10%,rgb(123,127,109) 10% 20%,rgb(140,139,132) 20% 30%,rgb(156,153,149) 30% 40%,rgb(164,161,155) 40% 50%,rgb(159,156,149) 50% 60%,rgb(141,140,133) 60% 70%,rgb(118,119,111) 70% 80%,rgb(101,105,89) 80% 90%,rgb(94,101,74) 90% 100%)',
      'linear-gradient(90deg,rgb(94,106,60) 10%,rgb(100,108,77) 10% 20%,rgb(113,116,98) 20% 30%,rgb(130,128,115) 30% 40%,rgb(140,136,123) 40% 50%,rgb(137,133,119) 50% 60%,rgb(120,118,104) 60% 70%,rgb(96,98,81) 70% 80%,rgb(76,84,56) 80% 90%,rgb(68,78,36) 90% 100%)',
      'linear-gradient(90deg,rgb(100,103,29) 10%,rgb(100,101,50) 10% 20%,rgb(107,104,72) 20% 30%,rgb(120,112,88) 30% 40%,rgb(130,118,95) 40% 50%,rgb(128,116,91) 50% 60%,rgb(114,104,76) 60% 70%,rgb(95,89,55) 70% 80%,rgb(82,79,31) 80% 90%,rgb(79,77,12) 90% 100%)',
      'linear-gradient(90deg,rgb(115,109,20) 10%,rgb(114,106,48) 10% 20%,rgb(117,106,71) 20% 30%,rgb(126,110,84) 30% 40%,rgb(133,114,87) 40% 50%,rgb(132,112,80) 50% 60%,rgb(121,102,65) 60% 70%,rgb(107,92,49) 70% 80%,rgb(99,88,36) 80% 90%,rgb(99,88,31) 90% 100%)',
    ].join(','),
    backgroundPosition: '0 0,0 11%,0 22%,0 33%,0 44%,0 56%,0 67%,0 78%,0 89%,0 100%',
    backgroundRepeat: 'no-repeat',
    backgroundSize: '100% 10%',
    filter: 'blur(24px)',
    transform: 'scale(1.2)',
  });
});


src/lib.rs
extern crate blurhash;
extern crate console_error_panic_hook;
extern crate serde;
extern crate serde_json;
extern crate wasm_bindgen;

use blurhash::decode;
use gloo_utils::format::JsValueSerdeExt;
use serde::{Deserialize, Serialize};
use std::panic;
use wasm_bindgen::prelude::*;

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BlurhashCss {
    background_image: String,
    background_position: String,
    background_repeat: String,
    background_size: String,
    filter: String,
    transform: String,
}

fn get_rounded_percentage_of(part: u32, whole: u32) -> u32 {
    let value = part as f32 / whole as f32 * 100 as f32;
    value.round() as u32
}

fn blurhash_to_bytes(hash: &String, width: u32, height: u32) -> Vec<u8> {
    let hash_slice: &str = &hash[..];
    let punch = 1.0;
    decode(hash_slice, width, height, punch)
}

fn get_pixel(pixel_bytes: &[u8], x: u32, y: u32, width: u32, index: u32) -> u8 {
    let number_of_channels = 4; // RGBa
    let bytes_per_row = width * number_of_channels;
    pixel_bytes[(number_of_channels * x + index + y * bytes_per_row) as usize]
}

fn blurhash_to_css_struct(hash: &String, width: u32, height: u32) -> BlurhashCss {
    panic::set_hook(Box::new(console_error_panic_hook::hook));
    let pixel_bytes = blurhash_to_bytes(&hash, width, height);
    let r_index = 0;
    let g_index = 1;
    let b_index = 2;
    let background_size = format!("100% {}%", 100.0 / height as f32);
    let mut background_position = Vec::new();
    let mut linear_gradients = Vec::new();

    for y in 0..height {
        let mut row_linear_gradients = Vec::new();

        for x in 0..width {
            let r = get_pixel(&pixel_bytes, x, y, width, r_index);
            let g = get_pixel(&pixel_bytes, x, y, width, g_index);
            let b = get_pixel(&pixel_bytes, x, y, width, b_index);

            let start_percent = if x == 0 {
                String::from("")
            } else {
                format!(" {}%", get_rounded_percentage_of(x, width))
            };

            let end_percent = if x == width {
                String::from("")
            } else {
                format!(" {}%", get_rounded_percentage_of(x + 1, width))
            };

            let linear_gradient = format!("rgb({},{},{}){}{}", r, g, b, start_percent, end_percent);
            row_linear_gradients.push(linear_gradient);
        }

        linear_gradients.push(format!(
            "linear-gradient(90deg,{})",
            row_linear_gradients.join(",")
        ));

        if y == 0 {
            background_position.push(String::from("0 0"));
        } else {
            background_position.push(format!("0 {}%", get_rounded_percentage_of(y, height - 1)));
        }
    }

    BlurhashCss {
        background_image: linear_gradients.join(","),
        background_position: background_position.join(","),
        background_repeat: String::from("no-repeat"),
        background_size: background_size,
        filter: String::from("blur(24px)"),
        transform: String::from("scale(1.2)"),
    }
}

#[wasm_bindgen]
pub fn blurhash_to_css(hash: String, width: u32, height: u32) -> String {
    let css_struct = blurhash_to_css_struct(&hash, width, height);
    serde_json::to_string(&css_struct).unwrap()
}

#[wasm_bindgen]
pub fn blurhashes_to_css(hashes: JsValue, width: u32, height: u32) -> String {
    let hashes: Vec<String> = hashes.into_serde().unwrap();
    let css_structs: Vec<BlurhashCss> = hashes
        .iter()
        .map(|hash| blurhash_to_css_struct(&hash, width, height))
        .collect();
    serde_json::to_string(&css_structs).unwrap()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn calculates_percentages() {
        assert_eq!(get_rounded_percentage_of(10, 100), 10);
        assert_eq!(get_rounded_percentage_of(25, 200), 13);
    }

    #[test]
    fn decodes_blurhash_into_bytes() {
        let hash = String::from("LEHLh[WB2yk8pyoJadR*.7kCMdnj");
        let width = 400;
        let height = 600;
        let pixel_bytes = blurhash_to_bytes(&hash, width, height);
        let expected_rgba_bytes = 960000;

        assert_eq!(pixel_bytes.len(), expected_rgba_bytes);
    }

    #[test]
    fn converts_blurhash_into_css_struct() {
        let hash = String::from("LEHLh[WB2yk8pyoJadR*.7kCMdnj");
        let width = 10;
        let height = 10;
        let css_struct = blurhash_to_css_struct(&hash, width, height);

        assert_eq!(
            css_struct.background_position,
            "0 0,0 11%,0 22%,0 33%,0 44%,0 56%,0 67%,0 78%,0 89%,0 100%"
        );
        assert_eq!(css_struct.background_repeat, "no-repeat");
        assert_eq!(css_struct.background_size, "100% 10%");
        assert_eq!(css_struct.filter, "blur(24px)");
        assert_eq!(css_struct.background_image, "linear-gradient(90deg,rgb(134,164,177) 10%,rgb(140,165,177) 10% 20%,rgb(153,170,177) 20% 30%,rgb(168,175,176) 30% 40%,rgb(178,179,173) 40% 50%,rgb(181,179,171) 50% 60%,rgb(176,177,170) 60% 70%,rgb(166,173,172) 70% 80%,rgb(152,169,176) 80% 90%,rgb(141,166,179) 90% 100%),linear-gradient(90deg,rgb(133,162,176) 10%,rgb(138,164,176) 10% 20%,rgb(152,168,175) 20% 30%,rgb(166,172,173) 30% 40%,rgb(176,175,170) 40% 50%,rgb(179,176,167) 50% 60%,rgb(174,174,167) 60% 70%,rgb(164,171,170) 70% 80%,rgb(151,167,174) 80% 90%,rgb(140,165,179) 90% 100%),linear-gradient(90deg,rgb(129,160,174) 10%,rgb(135,160,173) 10% 20%,rgb(148,162,169) 20% 30%,rgb(161,163,165) 30% 40%,rgb(171,164,159) 40% 50%,rgb(173,165,156) 50% 60%,rgb(169,165,157) 60% 70%,rgb(159,163,162) 70% 80%,rgb(147,162,170) 80% 90%,rgb(137,161,176) 90% 100%),linear-gradient(90deg,rgb(124,156,170) 10%,rgb(130,155,168) 10% 20%,rgb(142,153,162) 20% 30%,rgb(155,151,153) 30% 40%,rgb(164,149,144) 40% 50%,rgb(166,150,140) 50% 60%,rgb(162,151,142) 60% 70%,rgb(153,153,152) 70% 80%,rgb(142,155,163) 80% 90%,rgb(133,156,172) 90% 100%),linear-gradient(90deg,rgb(121,151,166) 10%,rgb(126,149,163) 10% 20%,rgb(137,144,153) 20% 30%,rgb(149,138,140) 30% 40%,rgb(158,134,127) 40% 50%,rgb(160,134,121) 50% 60%,rgb(156,138,127) 60% 70%,rgb(148,143,140) 70% 80%,rgb(138,148,156) 80% 90%,rgb(129,151,167) 90% 100%),linear-gradient(90deg,rgb(119,147,161) 10%,rgb(124,144,157) 10% 20%,rgb(135,137,146) 20% 30%,rgb(146,129,129) 30% 40%,rgb(155,124,114) 40% 50%,rgb(158,124,107) 50% 60%,rgb(154,130,114) 60% 70%,rgb(146,136,131) 70% 80%,rgb(136,142,149) 80% 90%,rgb(128,146,162) 90% 100%),linear-gradient(90deg,rgb(121,144,157) 10%,rgb(125,142,152) 10% 20%,rgb(135,135,141) 20% 30%,rgb(147,128,124) 30% 40%,rgb(156,124,108) 40% 50%,rgb(159,124,101) 50% 60%,rgb(157,129,108) 60% 70%,rgb(149,135,126) 70% 80%,rgb(138,140,143) 80% 90%,rgb(129,143,156) 90% 100%),linear-gradient(90deg,rgb(125,143,152) 10%,rgb(129,141,149) 10% 20%,rgb(139,137,139) 20% 30%,rgb(151,133,124) 30% 40%,rgb(160,131,111) 40% 50%,rgb(165,133,105) 50% 60%,rgb(163,136,111) 60% 70%,rgb(155,139,125) 70% 80%,rgb(143,141,140) 80% 90%,rgb(131,142,151) 90% 100%),linear-gradient(90deg,rgb(130,143,149) 10%,rgb(134,142,146) 10% 20%,rgb(144,142,139) 20% 30%,rgb(156,141,128) 30% 40%,rgb(166,142,119) 40% 50%,rgb(172,144,115) 50% 60%,rgb(170,146,118) 60% 70%,rgb(162,146,128) 70% 80%,rgb(148,144,139) 80% 90%,rgb(134,142,148) 90% 100%),linear-gradient(90deg,rgb(134,143,147) 10%,rgb(138,144,145) 10% 20%,rgb(147,145,139) 20% 30%,rgb(160,148,132) 30% 40%,rgb(171,151,125) 40% 50%,rgb(177,153,122) 50% 60%,rgb(176,153,125) 60% 70%,rgb(167,151,131) 70% 80%,rgb(152,147,139) 80% 90%,rgb(137,143,145) 90% 100%)");
        assert_eq!(css_struct.transform, "scale(1.2)");
    }

    #[test]
    fn converts_blurhash_into_css_json() {
        let hash = String::from("LEHLh[WB2yk8pyoJadR*.7kCMdnj");
        let width = 10;
        let height = 10;
        let css_json = blurhash_to_css(hash, width, height);

        assert_eq!(css_json, "{\"backgroundImage\":\"linear-gradient(90deg,rgb(134,164,177) 10%,rgb(140,165,177) 10% 20%,rgb(153,170,177) 20% 30%,rgb(168,175,176) 30% 40%,rgb(178,179,173) 40% 50%,rgb(181,179,171) 50% 60%,rgb(176,177,170) 60% 70%,rgb(166,173,172) 70% 80%,rgb(152,169,176) 80% 90%,rgb(141,166,179) 90% 100%),linear-gradient(90deg,rgb(133,162,176) 10%,rgb(138,164,176) 10% 20%,rgb(152,168,175) 20% 30%,rgb(166,172,173) 30% 40%,rgb(176,175,170) 40% 50%,rgb(179,176,167) 50% 60%,rgb(174,174,167) 60% 70%,rgb(164,171,170) 70% 80%,rgb(151,167,174) 80% 90%,rgb(140,165,179) 90% 100%),linear-gradient(90deg,rgb(129,160,174) 10%,rgb(135,160,173) 10% 20%,rgb(148,162,169) 20% 30%,rgb(161,163,165) 30% 40%,rgb(171,164,159) 40% 50%,rgb(173,165,156) 50% 60%,rgb(169,165,157) 60% 70%,rgb(159,163,162) 70% 80%,rgb(147,162,170) 80% 90%,rgb(137,161,176) 90% 100%),linear-gradient(90deg,rgb(124,156,170) 10%,rgb(130,155,168) 10% 20%,rgb(142,153,162) 20% 30%,rgb(155,151,153) 30% 40%,rgb(164,149,144) 40% 50%,rgb(166,150,140) 50% 60%,rgb(162,151,142) 60% 70%,rgb(153,153,152) 70% 80%,rgb(142,155,163) 80% 90%,rgb(133,156,172) 90% 100%),linear-gradient(90deg,rgb(121,151,166) 10%,rgb(126,149,163) 10% 20%,rgb(137,144,153) 20% 30%,rgb(149,138,140) 30% 40%,rgb(158,134,127) 40% 50%,rgb(160,134,121) 50% 60%,rgb(156,138,127) 60% 70%,rgb(148,143,140) 70% 80%,rgb(138,148,156) 80% 90%,rgb(129,151,167) 90% 100%),linear-gradient(90deg,rgb(119,147,161) 10%,rgb(124,144,157) 10% 20%,rgb(135,137,146) 20% 30%,rgb(146,129,129) 30% 40%,rgb(155,124,114) 40% 50%,rgb(158,124,107) 50% 60%,rgb(154,130,114) 60% 70%,rgb(146,136,131) 70% 80%,rgb(136,142,149) 80% 90%,rgb(128,146,162) 90% 100%),linear-gradient(90deg,rgb(121,144,157) 10%,rgb(125,142,152) 10% 20%,rgb(135,135,141) 20% 30%,rgb(147,128,124) 30% 40%,rgb(156,124,108) 40% 50%,rgb(159,124,101) 50% 60%,rgb(157,129,108) 60% 70%,rgb(149,135,126) 70% 80%,rgb(138,140,143) 80% 90%,rgb(129,143,156) 90% 100%),linear-gradient(90deg,rgb(125,143,152) 10%,rgb(129,141,149) 10% 20%,rgb(139,137,139) 20% 30%,rgb(151,133,124) 30% 40%,rgb(160,131,111) 40% 50%,rgb(165,133,105) 50% 60%,rgb(163,136,111) 60% 70%,rgb(155,139,125) 70% 80%,rgb(143,141,140) 80% 90%,rgb(131,142,151) 90% 100%),linear-gradient(90deg,rgb(130,143,149) 10%,rgb(134,142,146) 10% 20%,rgb(144,142,139) 20% 30%,rgb(156,141,128) 30% 40%,rgb(166,142,119) 40% 50%,rgb(172,144,115) 50% 60%,rgb(170,146,118) 60% 70%,rgb(162,146,128) 70% 80%,rgb(148,144,139) 80% 90%,rgb(134,142,148) 90% 100%),linear-gradient(90deg,rgb(134,143,147) 10%,rgb(138,144,145) 10% 20%,rgb(147,145,139) 20% 30%,rgb(160,148,132) 30% 40%,rgb(171,151,125) 40% 50%,rgb(177,153,122) 50% 60%,rgb(176,153,125) 60% 70%,rgb(167,151,131) 70% 80%,rgb(152,147,139) 80% 90%,rgb(137,143,145) 90% 100%)\",\"backgroundPosition\":\"0 0,0 11%,0 22%,0 33%,0 44%,0 56%,0 67%,0 78%,0 89%,0 100%\",\"backgroundRepeat\":\"no-repeat\",\"backgroundSize\":\"100% 10%\",\"filter\":\"blur(24px)\",\"transform\":\"scale(1.2)\"}");
    }
}
